lib/manageiq/release/mirror_helper.rb

```diff
@@ -0,0 +1,175 @@
+      end
+
+      def mirror(repo, remote_source)
+        with_repo(repo, remote_source) do
+          send("mirror_#{remote_source}_repo", repo)
+        end
+      end
+
+      private
+
+      def mirror_branches_for(repo)
+        Settings.branch_mirror_defaults.to_h.merge(Settings.branch_mirror_overrides[repo].to_h || {}).each_with_object({}) { |(k, v), h| h[k.to_s] = v }
+      end
+
+      def mirror_branches(repo, source_remote, dest_remote)
+        mirror_branches_for(repo).each do |source_name, dest_name|
+          sync_branch(source_remote, source_name, dest_remote, dest_name)
+        end
+      end
+
+      def mirror_upstream_repo(repo)
+        mirror_remote_refs("upstream", "downstream")
+        mirror_branches(repo, "upstream", "downstream")
+        mirror_remote_refs("downstream", "backup")
+      end
+
+      def mirror_downstream_repo(repo)
+        mirror_branches(repo, "downstream", "downstream")
+        mirror_remote_refs("downstream", "backup")
+      end
+
+      def dry_run?
+        return @dry_run if defined?(@dry_run)
+        @dry_run = ARGV.include?("--dry-run")
+      end
+
+      def downstream_repo_name(repo)
+        repo.sub("manageiq", Settings.productization_name)
+      end
+
+      def system(*args)
+        puts "+ #{"dry_run: " if dry_run?}#{args.join(" ")}"
+        return true if dry_run?
+
+        args << {} unless args.last.is_a?(Hash)
+        args.last[[:out, :err]] = ["/tmp/mirror_helper_out", "w"]
+
+        super.tap do |result|
+          puts "!!! An error has occurred:\n#{File.read("/tmp/mirror_helper_out")}".bold.red unless result
+        end
+      end
+
+      def with_repo(repo, remote_source)
+        repo_name = downstream_repo_name(repo)
+        puts "\n==== Mirroring #{repo_name} ====".bold.cyan
+
+        path = "#{Settings.working_directory}/#{repo_name}"
+        clone_repo(repo, repo_name, path, remote_source) unless File.directory?(path)
+
+        Dir.chdir(path) do
+          puts "\n==== Fetching for #{repo_name} ====".bold.green
+          # Enforce an order for remote fetching to ensure that moved
+          #   tags prefer what is on upstream
+          system("git fetch backup --prune --tags") if remote_exists?("backup")
+          system("git fetch downstream --prune --tags")
+          system("git fetch upstream --prune --tags") if [:red_hat_cloudforms, :upstream].include?(remote_source)
+
+          yield
+        end
+
+        puts
+      end
+
+      def clone_repo(upstream_repo, downstream_repo, path, remote_source)
+        system("git clone #{Settings.remotes[remote_source]}/#{upstream_repo}.git #{path} -o upstream")
+        Dir.chdir(path) do
+          system("git remote add downstream #{Settings.remotes.downstream}/#{downstream_repo}.git") unless remote_exists?("downstream")
+          system("git remote add backup #{Settings.remotes.backup}/#{downstream_repo}.git")
```

Comment:
If you don't have a backup, then this adds a weird remote.  This should probably be protected with an `if Settings.remotes.backup`

Reply:
or, since there's a double-check on downstream, we probably needs something here as well, so...

```ruby
if Settings.remotes.backup && !remote_exists?("backup")
```

Reply:
Done